src/Microsoft.OData.Core/UriParser/Binders/MetadataBindingUtils.cs

```diff
@@ -58,19 +58,31 @@ internal static SingleValueNode ConvertToTypeIfNeeded(SingleValueNode source, IE
                 }
 
                 ConstantNode constantNode = source as ConstantNode;
-                if (constantNode != null && constantNode.Value != null && source.TypeReference.IsString() && targetTypeReference.IsEnum())
+                // Check if the source node is a constant node, not null, and the source type is either string or integral
+                // and the target type is an enum.
+                if (constantNode != null && constantNode.Value != null && (source.TypeReference.IsString() || source.TypeReference.IsIntegral()) && targetTypeReference.IsEnum())
                 {
                     string memberName = constantNode.Value.ToString();
                     IEdmEnumType enumType = targetTypeReference.Definition as IEdmEnumType;
                     if (enumType.Members.Any(m => string.Compare(m.Name, memberName, StringComparison.Ordinal) == 0))
                     {
                         string literalText = ODataUriUtils.ConvertToUriLiteral(constantNode.Value, default(ODataVersion));
-                        return new ConstantNode(new ODataEnumValue(constantNode.Value.ToString(), targetTypeReference.Definition.ToString()), literalText, targetTypeReference);
+                        return new ConstantNode(new ODataEnumValue(memberName, enumType.ToString()), literalText, targetTypeReference);
                     }
-                    else
+
+                    // If the member name is an integral value, we should try to convert it to the enum member name.
+                    if (int.TryParse(memberName, out int memberIntegralValue))
                     {
-                        throw new ODataException(ODataErrorStrings.Binder_IsNotValidEnumConstant(memberName));
+                        // Find the enum member with the matching integral value
+                        var enumMember = enumType.Members.FirstOrDefault(m => m.Value.Value == memberIntegralValue);

```

Comment:
This predicate passed to `FirstOrDefault` will cause allocations because of capturing the `memberIntegralValue` variable.

Reply:
@habbes 

Thank you for pointing this perf issue. I have added a public method in [EnumHelper](https://github.com/OData/odata.net/blob/fix/2542-filter-with-enum-integer/src/Microsoft.OData.Edm/ExtensionMethods/EnumHelper.cs) called [bool TryParseEnum(this IEdmEnumType enumType, long value, out IEdmEnumMember enumMember)](https://github.com/OData/odata.net/blob/fix/2542-filter-with-enum-integer/src/Microsoft.OData.Edm/ExtensionMethods/EnumHelper.cs#L136) that finds enum member value using enum integral value. I then used this method here like so:

```cs
if (enumType.TryParseEnum(memberIntegralValue, out IEdmEnumMember enumMember))
{
    string literalText = ODataUriUtils.ConvertToUriLiteral(enumMember.Name, default(ODataVersion));
    ...
}
```

Reply:
@habbes

I have also added a method in [EnumHelper](https://github.com/OData/odata.net/blob/fix/2542-filter-with-enum-integer/src/Microsoft.OData.Edm/ExtensionMethods/EnumHelper.cs) called [bool EnumMemberExists(this IEdmEnumType enumType, string memberName, StringComparison comparison = StringComparison.Ordinal)](https://github.com/OData/odata.net/blob/fix/2542-filter-with-enum-integer/src/Microsoft.OData.Edm/ExtensionMethods/EnumHelper.cs#L158) that checks if the given member name exists in the enum type.

This is used in the place where it was using `.Any`
```cs
if(enumType.EnumMemberExists(memberName, StringComparison.Ordinal))
{
    string literalText = ODataUriUtils.ConvertToUriLiteral(constantNode.Value, default(ODataVersion));
   ...
}
```